src/Disciplina/Crypto/MerkleTree.hs

```diff
@@ -0,0 +1,232 @@
+                 -> MerkleSignature a
+mkBranchRootHash (MerkleSignature (AbstractHash hl) sl)
+                 (MerkleSignature (AbstractHash hr) sr)
+  = MerkleSignature
+     (hash $ toLazyByteString $ mconcat
+        [ byteString (one sl)
+        , byteString (one sr)
+        , byteString (convert hl)
+        , byteString (convert hr) ])
+     (sl + sr)
+  where
+    toLazyByteString :: Builder -> LBS.ByteString
+    toLazyByteString = Builder.toLazyByteStringWith (Builder.safeStrategy 1024 4096) mempty
+
+-- | Smart constructor for MerkleTree.
+fromFoldable :: (HasHash a, Foldable t) => t a -> MerkleTree a
+fromFoldable = fromList . F.toList
+
+-- | Smart constructor for MerkleTree.
+fromContainer :: (HasHash (Element t), Container t) => t -> MerkleTree (Element t)
+fromContainer = fromFoldable . toList
+
+fromList :: HasHash a => [a] -> MerkleTree a
+fromList [] = MerkleEmpty
+fromList ls = MerkleTree (go' 1 (length lsIndexed))
+  where
+    lsIndexed = zip [1,2..] ls
+    arr = array (1, length lsIndexed) lsIndexed
+    go' lo hi = if lo == hi
+                then mkLeaf (lo - 1, arr ! lo)
+                else mkBranch (go' loL hiL) (go' loR hiR)
+      where
+        -- new borders
+        (!loL, !hiL, !loR, !hiR) = (lo, lo + i, lo + i + 1, hi)
+        i = powerOfTwo (hi - lo)
+
+-- | Returns root of merkle tree.
+getMerkleRoot :: MerkleTree a -> MerkleSignature a
+getMerkleRoot MerkleEmpty    = emptyHash
+getMerkleRoot (MerkleTree x) = mRoot x
+
+emptyHash :: MerkleSignature a
+emptyHash = MerkleSignature (hash mempty) 0
+
+-- | Return the largest power of two such that it's smaller than X.
+--
+-- >>> powerOfTwo 64
+-- 32
+-- >>> powerOfTwo 65
+-- 64
+powerOfTwo :: (Bits a, Num a) => a -> a
+powerOfTwo n
+    | n .&. (n - 1) == 0 = n `shiftR` 1
+    | otherwise = go n
+ where
+    {- “x .&. (x - 1)” clears the least significant bit:
+           ↓
+       01101000     x
+       01100111     x - 1
+       --------
+       01100000     x .&. (x - 1)
+     -}
+    go w = if w .&. (w - 1) == 0 then w else go (w .&. (w - 1))
+
+data MerkleProof a = MerkleProof
+  { getMerkleProof     :: [ProofElem a]  -- ^ list of proof elements
+  , getMerkleProofRoot :: Maybe (MerkleSignature a) -- ^ leaf root proof is constructed for
+  }
+  deriving (Show, Eq, Ord, Generic, Serialise)
+
+data ProofElem a = ProofElem
+  { nodeRoot    :: MerkleSignature a
+  , siblingRoot :: MerkleSignature a
+  , nodeSide    :: Side
+  } deriving (Show, Eq, Ord, Generic, Serialise)
+
+data Side = L | R
+  deriving (Show, Eq, Ord, Generic, Serialise)
+
+-- | Construct merkle proof by recusive walking the tree and collecting ProofElem until we hit
+-- matching leafRoot.
+mkMerkleProof :: forall a. MerkleTree a -- ^ merkle tree we want to construct a proof from
+                        -> LeafIndex    -- ^ leaf index used for proof
+                        -> MerkleProof a
+mkMerkleProof MerkleEmpty _ = MerkleProof [] Nothing
+mkMerkleProof (MerkleTree rootNode) n =
```

Comment:
In the task requirements there was also options of providing a set or range of indices and getting the _batch proof_ (i. e. a single subgraph of a Merkle tree including all required leaves). This feature is important, because it's required for preparing a data response for data disclosure.
So I propose the following:
- Change the type for proof (make it not a list of nodes with hanging siblings, but an actual tree with some internal nodes allowed to have no children)
- Make a generic method for preparing a proof for _set_ of indices (let's omit ranges for now), and implement method for a single index as a special case.